src/DispatchFunctional.jl

```diff
@@ -0,0 +1,155 @@
+using DftFunctionals
+import Libxc
+
+#
+# Libxc (TODO Move this upstream, changing the interface of Libxc.jl)
+#
+struct LibxcFunctional{Family,Kind} <: Functional{Family,Kind}
+    identifier::Symbol
+end
+function LibxcFunctional(identifier::Symbol)
+    fun = Libxc.Functional(identifier)
+    @assert fun.kind   in (:exchange, :correlation, :exchange_correlation)
+    kind = Dict(:exchange => :x, :correlation => :c, :exchange_correlation => :xc)[fun.kind]
+
+    @assert fun.family in (:lda, :gga, :mgga)  # Hybrids not supported yet.
+    if fun.family == :mgga && Libxc.needs_laplacian(fun)
+        family = :mggal
+    else
+        family = fun.family
+    end
+    LibxcFunctional{family,kind}(identifier)
+end
+
+DftFunctionals.identifier(fun::LibxcFunctional) = fun.identifier
+function DftFunctionals.has_energy(func::LibxcFunctional)
+    0 in Libxc.supported_derivatives(Libxc.Functional(func.identifier))
+end
+
+function libxc_unfold_spin(data::Matrix, n_spin::Int)
+    n_p = size(data, 2)
+    if n_spin == 1
+        data  # Only one spin component
+    elseif n_spin == 2
+        unfolded = similar(data, 2, 2, n_p)
+        unfolded[1, 1, :] = data[1, :]
+        unfolded[2, 2, :] = data[3, :]
+
+        unfolded[1, 2, :] = unfolded[2, 1, :] = data[2, :]
+        unfolded
+    elseif n_spin == 3
+        unfolded = similar(data, 3, 3, n_p)
+        unfolded[1, 1, :] = data[1, :]
+        unfolded[2, 2, :] = data[4, :]
+        unfolded[3, 3, :] = data[6, :]
+
+        unfolded[1, 2, :] = unfolded[2, 1, :] = data[2, :]
+        unfolded[1, 3, :] = unfolded[3, 1, :] = data[3, :]
+        unfolded[2, 3, :] = unfolded[3, 2, :] = data[5, :]
+        unfolded
+    else
+        error("Unspoorted n_spin == $n_spin")
+    end
+end
+
+libxc_energy(terms, ρ) = haskey(terms, :zk) ? reshape(terms.zk, 1, size(ρ, 2)) .* ρ : false
+
+function DftFunctionals.potential_terms(func::LibxcFunctional{:lda}, ρ::Matrix{Float64})
+    s_ρ, n_p = size(ρ)
+    fun = Libxc.Functional(func.identifier; n_spin=s_ρ)
+    derivatives = filter(i -> i in Libxc.supported_derivatives(fun), 0:1)
+    terms = Libxc.evaluate(fun; rho=ρ, derivatives)
+    e  = libxc_energy(terms, ρ)
+    Vρ = reshape(terms.vrho, s_ρ, n_p)
+    (; e, Vρ)
+end
+function DftFunctionals.potential_terms(func::LibxcFunctional{:gga}, ρ::Matrix{Float64},
+                                        σ::Matrix{Float64})
+    s_ρ, n_p = size(ρ)
+    s_σ = size(σ, 1)
+    fun = Libxc.Functional(func.identifier; n_spin=s_ρ)
+    derivatives = filter(i -> i in Libxc.supported_derivatives(fun), 0:1)
+    terms = Libxc.evaluate(fun; rho=ρ, sigma=σ, derivatives)
```

Comment:
that's a lot of copy/paste, can't you make an internal function or something here?

Reply:
Don't think that's worth it. The old Libxc interface will die anyway.